File: src/core/orphan-matcher.ts
// Orphan Matcher — Pure functions for orphan page linking
// Extracted from lint-fixes.ts::linkOrphanPage()
// Zero side effects, fully testable

export interface OrphanContext {
  orphanContent: string;
  wikiIndex: string;
}

export interface OrphanLinkSuggestion {
  pagePath: string;
  linkText: string;
  linkTarget: string;
}

/**
 * Build prompt for linking orphan pages to related content.
 *
 * @param template - The prompt template with {{placeholders}}
 * @param context - Context information for the orphan page
 * @returns Constructed prompt ready for LLM
 */
export function buildOrphanLinkPrompt(
  template: string,
  context: OrphanContext
): string {
  // Truncate content and index to prevent token overflow
  const truncatedContent = context.orphanContent.substring(0, 2000);
  const truncatedIndex = context.wikiIndex.substring(0, 3000);

  return template
    .replace(/\{\{orphan_content\}\}/g, truncatedContent)
    .replace(/\{\{wiki_index\}\}/g, truncatedIndex);
}

/**
 * Validate if a suggested orphan link is valid (target exists in content).
 *
 * @param relatedContent - Content of the related page
 * @param linkTarget - The link target to validate
 * @returns True if target exists in content
 */
export function validateOrphanLinkTarget(
  relatedContent: string,
  linkTarget: string
): boolean {
  return relatedContent.includes(linkTarget);
}

/**
 * Build the content update for linking an orphan page.
 *
 * @param relatedContent - Current content of related page
 * @param suggestion - Link suggestion from LLM
 * @param sectionHeader - Header for related pages section
 * @returns Updated content with orphan link added
 */
export function buildOrphanLinkUpdate(
  relatedContent: string,
  suggestion: OrphanLinkSuggestion,
  sectionHeader: string
): string {
  const header = `## ${sectionHeader}`;
  const section = relatedContent.includes(header) ? '' : `\n\n${header}`;
  return `${relatedContent}${section}\n- ${suggestion.linkText} ${suggestion.linkTarget}`;
}

/**
 * Normalize page path to full path.
 *
 * @param pagePath - Page path (may be relative or full)
 * @param wikiFolder - Wiki root folder
 * @returns Full path
 */
export function normalizeOrphanPagePath(
  pagePath: string,
  wikiFolder: string
): string {
  return pagePath.startsWith(wikiFolder)
    ? pagePath
    : `${wikiFolder}/${pagePath}`;
}

File: python_lib/llm_wiki/wiki/lint/orphan_matcher.py
"""
Orphan-matcher helpers — port of src/core/orphan-matcher.ts
"""
from __future__ import annotations

from typing import Dict, Optional


def build_orphan_link_prompt(template: str, orphan_content: str, wiki_index: str, wiki_folder: str) -> str:
    return (
        template
        .replace("{{orphan_content}}", orphan_content[:2000])
        .replace("{{wiki_index}}", wiki_index[:3000])
        .replace("{{wikiFolder}}", wiki_folder)
    )


def validate_orphan_link_target(related_content: str, link_target: str) -> bool:
    return link_target in related_content


def build_orphan_link_update(
    related_content: str,
    page_path: str,
    link_text: str,
    link_target: str,
    section_header: str,
) -> str:
    header = f"## {section_header}"
    section = "" if header in related_content else f"\n\n{header}"
    return f"{related_content}{section}\n- {link_text} {link_target}"


def normalize_orphan_page_path(page_path: str, wiki_folder: str) -> str:
    return page_path if page_path.startswith(wiki_folder) else f"{wiki_folder}/{page_path}"
